crates/amaru-kernel/src/protocol_parameters.rs

```diff
@@ -22,38 +31,393 @@ pub struct ProtocolParameters {
     pub ref_script_cost_stride: u32,
     pub ref_script_cost_multiplier: RationalNumber,
 
-    pub max_epoch: Epoch,
-    pub pool_influence: RationalNumber,
-    pub desired_pool_count: u32,
-    pub treasury_expansion_rate: RationalNumber,
-    pub monetary_expansion_rate: RationalNumber,
-    pub collateral_percentage: u32,
+    // Technical group
+    pub max_epoch: EpochInterval,
+    pub optimal_stake_pools_count: u16,
+    pub pledge_influence: RationalNumber,
+    pub collateral_percentage: u16,
     pub cost_models: CostModels,
 
+    // Governance group
     pub pool_thresholds: PoolThresholds,
     pub drep_thresholds: DrepThresholds,
-    pub cc_min_size: u32,
-    pub cc_max_term_length: u32,
-    pub gov_action_lifetime: u32,
+    pub cc_min_size: u16,
+    pub cc_max_term_length: EpochInterval,
+    pub gov_action_lifetime: EpochInterval,
     pub gov_action_deposit: Coin,
     pub drep_deposit: Coin,
-    pub drep_activity: Epoch,
+    pub drep_expiry: Epoch,
+}
+
+fn decode_u16(d: &mut Decoder<'_>) -> Result<u16, cbor::decode::Error> {
+    #[allow(clippy::wildcard_enum_match_arm)]
+    match d.datatype()? {
+        cbor::data::Type::U8 => Ok(d.u8()? as u16),
+        cbor::data::Type::U16 => Ok(d.u16()?),
+        _ => Err(cbor::decode::Error::message("Expected u8/u16 type")),
+    }
 }
 
-#[derive(Debug, Clone)]
+fn decode_u32(d: &mut Decoder<'_>) -> Result<u32, cbor::decode::Error> {
+    #[allow(clippy::wildcard_enum_match_arm)]
+    match d.datatype()? {
+        cbor::data::Type::U8 => Ok(d.u8()? as u32),
+        cbor::data::Type::U16 => Ok(d.u16()? as u32),
+        cbor::data::Type::U32 => Ok(d.u32()?),
+        _ => Err(cbor::decode::Error::message("Expected u8/u16/u32 type")),
+    }
+}
+
+fn decode_u64(d: &mut Decoder<'_>) -> Result<u64, cbor::decode::Error> {
+    #[allow(clippy::wildcard_enum_match_arm)]
+    match d.datatype()? {
+        cbor::data::Type::U8 => Ok(d.u8()? as u64),
+        cbor::data::Type::U16 => Ok(d.u16()? as u64),
+        cbor::data::Type::U32 => Ok(d.u32()? as u64),
+        cbor::data::Type::U64 => Ok(d.u64()?),
+        _ => Err(cbor::decode::Error::message("Expected u8/u16/u32/u64 type")),
+    }
+}
+
+fn decode_rationale(d: &mut Decoder<'_>) -> Result<RationalNumber, cbor::decode::Error> {
+    d.tag()?;
+    d.array()?;
+    let numerator = decode_u64(d)?;
+    let denominator = decode_u64(d)?;
+    Ok(RationalNumber {
+        numerator,
+        denominator,
+    })
+}
+
+impl<'b, C> cbor::decode::Decode<'b, C> for ProtocolParameters {
```

Comment:
I feel like this structure is perhaps "mechanical enough" that we can leverage the macro annotation for writing the encoders/decoders (and reduce the boilerplate here)? Although, I think the downside is that we have to order fields in the Rust structure in the same way they are serialised (loosing the ability to nicely group them for readability). 

So unsure. I think I value more the Rust readability than reducing a bit of boilerplate (which is now already written anyway :D !)